lib/devices/ios/xcode.js

```diff
@@ -4,19 +4,62 @@ var logger = require('../../server/logger.js').get('appium')
   , fs = require('fs')
   , path = require('path')
   , async = require('async')
-  , exec = require('child_process').exec;
+  , exec = require('child_process').exec
+  , helpers = require('../../helpers.js')
+  , env = process.env;
 
+var XCODE_SUBDIR = "/Contents/Developer";
 var xcode = {};
 
+var _hasExpectedSubDir = function (path) {
+  return path.substring(path.length - XCODE_SUBDIR.length) === XCODE_SUBDIR;
+};
+
+var _getPathFromSymlink = function (stdout, cb) {
+  // Node's invocation of xcode-select sometimes flakes and returns an empty string.
+  // Not clear why. As a workaround, Appium can reliably deduce the version in use by checking
+  // the locations xcode-select uses to store the selected version's path. This should be 100%
+  // reliable so long as the link locations remain the same. However, since we're relying on
+  // hardcoded paths, this approach will break the next time Apple changes the symlink location.
+  var symlnkPath = "/var/db/xcode_select_link";
+  var legacySymlnkPath = "/usr/share/xcode-select/xcode_dir_path"; //  Xcode < 5.x (?)
+
+  // xcode-select allows users to override its settings with the DEVELOPER_DIR env var,
+  // so check that first
+  if (helpers.notNullEmptyOrUndefined(env.DEVELOPER_DIR)) {
+    var dir = _hasExpectedSubDir(env.DEVELOPER_DIR) ?
+                env.DEVELOPER_DIR :
+                env.DEVELOPER_DIR + XCODE_SUBDIR;
+    cb(null, dir);
+  } else if (fs.existsSync(symlnkPath)) {
+    helpers.readlink(symlnkPath, cb);
+  } else if (fs.existsSync(legacySymlnkPath)) {
+    helpers.readlink(legacySymlnkPath, cb);
+  } else {
+    // We should only get here is we failed to capture xcode-select's stdout and our
+    // other checks failed. Either Apple has moved the symlink to a new location or the user
+    // is using the default install. 99.999% chance it's the latter, so issue a warning
+    // should we ever hit the edge case.
+    logger.warn("xcode-select returned an invalid path: " + stdout +
+                ". No links to alternate versions were detected. Using Xcode's " +
+                "default location.");
+    cb(null, "/Applications/Xcode.app" + XCODE_SUBDIR);
+  }
+};
+
 xcode.getPath = function (cb) {
-  exec('xcode-select --print-path', { maxBuffer: 524288, timeout: 3000 }, function (err, stdout, stderr) {
+  exec('xcode-select -p', { maxBuffer: 524288, timeout: 3000 }, function (err, stdout, stderr) {
```

Comment:
does the `-p` shortcut work with older xcodes too? for some reason it feels newer to me